feat: sound effects system

marcosvf132 opened the pull request:
## Co-Author
@dudantas 
Client: https://github.com/dudantas/tibia-client/releases/tag/13.16.13209
## Old PR
[Old PR](https://github.com/opentibiabr/canary/pull/531) changes moved to this one to separate features by commits

# Preview
https://user-images.githubusercontent.com/66353315/193288117-be5d4df1-d6ac-4a07-8e6d-84cf00542327.mp4

# 13.10
![image](https://user-images.githubusercontent.com/8551443/208708297-63810753-eb63-4f09-9e73-71856bad6fb4.png)

# WIP
- This PR implement a huge amount of features ingame that need time and test to verify all of then and to adjust the details on it.

# Musics and anthem
- Musics and anthem are not on this PR yet because i'm still catalogating then all, check for further commits on this PR.

# Features implemented so far

## Player combat
- Player weapon sounds are implemented, with future adjustments like a specifc sound on a different item.
- Wand and rods so far have the same sound, which will be changed on the next commits.

## Monster combat
- There is a script that run on monster spell declarations that automaticlly detect which spell sound it should be.
- If the user want's to select a custom sound to a spell, can use the following tags:
![image](https://user-images.githubusercontent.com/66353315/193341050-79fd423c-ebcb-4ae3-9c66-4071d66216b3.png)
![image](https://user-images.githubusercontent.com/66353315/193341073-44f95b87-b5bc-459e-87c7-e8252dcd7500.png)


## Item movement
- All items have sound on their movement, some can have the wrong sound, if you know an item that does not have the sound so accurated, please help me improving it.

## Npc sounds
- The NPC Canary does emit sounds on his sourrounds, check the new functions label on this description for more information.

# New function

# Enums
- We have  a LOT of sounds on tibia (512 so far), all those sounds are catalogated on the **/src/creatures/creatures_definitions.hpp** file, rigth here:
![image](https://user-images.githubusercontent.com/66353315/193290273-08f5ef60-5199-4fa7-b184-5aea27b6a77d.png)
###       NOTE: You can use then all on LUA env


# C++

## Sound movement
- The item sound movement comes from this new function, based on item loot category. I know this is not 100% accurated but its the best shot so far.
- The sound can change if the it's is being moved inside depot or player backpacks/inventory on the game.cpp call.
![image](https://user-images.githubusercontent.com/66353315/193290783-600baa7f-3e66-4992-9d1b-6e7fef4a3e17.png)

## Sound notification to spectators
- On this topic we have 2 new functions, the first one is used to send a unique sound to the spectators on the area. It already check the sound source if it's coming from the own player, other player, a creature or it's meant to be a global sound. This source type is used on the client to balance and mix the volume as the player desire.
![image](https://user-images.githubusercontent.com/66353315/193291336-5b8b305a-9b26-465d-a48e-96d0e326c3c3.png)
- This second function is used on distance weapons or any other sound that must be reproduced on a sequence. Like a casting sound and an impact sound.
![image](https://user-images.githubusercontent.com/66353315/193291575-ff582159-1489-4cca-8306-198ed5fbe728.png)

## OnThink sounds
- Some creatures emits sounds while in battle or when you get close to then. The first function is for the NPCs:
![image](https://user-images.githubusercontent.com/66353315/193292147-123c16df-c2af-4364-95f8-5944263a4c8a.png)
- The monsters sounds:
![image](https://user-images.githubusercontent.com/66353315/193292296-3c6eb5bb-1a25-410a-90d1-a37d5239a56f.png)

## Player client functions
- The player.h functions are used to send the sound to the client, can be used to send a sound directly to the player client and not send to all the players on the area.
![image](https://user-images.githubusercontent.com/66353315/193293086-516315f5-65a2-49ad-b299-5b01e186fe29.png)

## Player sound functions
- Functions that return the sound ID by the weapond attack and impact sound.
- Impact function:
![image](https://user-images.githubusercontent.com/66353315/193293341-70431b29-e682-4971-999b-eaf8ca23fdd3.png)
- Attack function:
![image](https://user-images.githubusercontent.com/66353315/193293410-e654590b-6562-4d19-8705-fb8a3e6b3505.png)

# LUA

## Spell (local spell = Spell("spell"))
- We have the possibility to add sounds on the spell, can be the sound when casting the spell AND the sound when the spell hits the target/area.
- Casting sound:
![image](https://user-images.githubusercontent.com/66353315/193293797-008812fb-dda3-47cf-90de-151ef984fc22.png)
- Impact sound:
![image](https://user-images.githubusercontent.com/66353315/193293854-9b222cd4-e237-4e63-923d-c6479534b684.png)

## Bestiary charm
- We have the possibility to add sounds on the spell, can be the sound when casting the spell AND the sound when the spell hits the target/area.
![image](https://user-images.githubusercontent.com/66353315/193294228-83387fa1-6ef4-4357-a3c7-4284ca550ee4.png)

## Npc
- You can add sound effects on the NPC just like the old text system, can be used as a environment sound as a blacksmith sound, hitting a wood, casting a spell etc.
- A random sound will be picked on the list to be reproduced.
![image](https://user-images.githubusercontent.com/66353315/193294768-4eaa436a-b8d2-4802-8803-5ac71a8347c6.png)

## Position
- You can used this function to send sounds on the specific position, just like the position:sendMagicEffect(...)
![image](https://user-images.githubusercontent.com/66353315/193295067-4964bb5b-ee21-4667-920d-909fb0f945f7.png)
![image](https://user-images.githubusercontent.com/66353315/193295095-5d706feb-15b9-43bb-984c-a2cd7d279ae0.png)
### The optional 'actor' on this function right here can be used to specify if you want to send the sound pretending is was coming from the creature. If you don't send a creature on the param, it will be sent as a global sound.

# XML
- The imbuement system can emits sounds too, you can set a sound on the imbuements.xml files be using the id:
![image](https://user-images.githubusercontent.com/66353315/193295564-6098dba7-f309-4ad2-be44-b3255c621009.png)

dudantas committed:
feat: sound effects system (source code and lua core)

The sound system will be split into two commits, one for the core (required) and the other for the scripts (optional).

This commit contains the "core" part of source and lua, they are needed for the scripts with sound modifications to work correctly and are needed for the server and the sounds to work correctly.

vitorfrancisco commented:
13.11 ou 13.10 pra fazer funcionar? to tomando crash no 13.11

elsongabriel commented:
> 13.11 ou 13.10 pra fazer funcionar? to tomando crash no 13.11

então mano. teve uma atualização na main, que mudou a versão 12.91 para 13.16.
consequentemente o client 13.11 e 13.10, não vai mais permitir logar-se.

o PR anterior #531 foi fechado, e novos PR's com as funcionalidades do client 13 foram abertos.

LeoBorges18 commented:
> ![image](https://user-images.githubusercontent.com/82473579/234477600-a0859d52-462d-4f43-9b35-cea7ab9e2d58.png) bosstiary problem of not appearing monster skin.

This problem has nothing to do with this pr. It will probably be appreances.dat from the client or datapack.

dudantas commented:
> 13.11 ou 13.10 pra fazer funcionar? to tomando crash no 13.11

13.16

Issolofreedom commented:
Sons dos potions não estão saindo.

Issolofreedom commented:
NPC Chuckles em yalahar fazendo o client crashar

QuebradaZN commented:
> > NPC Chuckles em yalahar fazendo o client crashar
> 
> Testei lá e aqui não aconteceu.

Aqui tambem, testei e não crashou o Client!

LeoBorges18 commented:
> NPC Chuckles em yalahar fazendo o client crashar

Testei aqui e não crashou tb

dudantas commented:
> Sons dos potions não estão saindo.

Isto vai ficar para depois.. Aparentemente o id do som está errado, se alguém souber o id me avisa.

QuebradaZN commented:
> > Sons dos potions não estão saindo.
> 
> Isto vai ficar para depois.. Aparentemente o id do som está errado, se alguém souber o id me avisa.

Dos Foods Tambem não!

dudantas committed:
feat: sound system lua scripts (depends of source and core lua codes)

This commit contains the non-mandatory files for the sound system, which are not absolutely necessary, but if not added, certain sounds will not work. To see the commit where the "core" part was added, look at the commit before this one.

elsongabriel commented:
testado.. e agora tá ok..

Lanemosme commented:
Pessoal, andei fazendo uns testes aqui, e o cliente tá crashando quando usado uma mana potion, e algo bem aleatório, as vezes crasha e as vezes não, mais alguém passando por isso?

dudantas commented:
> Pessoal, andei fazendo uns testes aqui, e o cliente tá crashando quando usado uma mana potion, e algo bem aleatório, as vezes crasha e as vezes não, mais alguém passando por isso?

Aqui não tive isso. Não sei se faz sentido também, porque a alteração que teve foi o som, aí daria debug sempre. Você ta usando a branch limpa? Client 13.16?

Lanemosme commented:
> > Pessoal, andei fazendo uns testes aqui, e o cliente tá crashando quando usado uma mana potion, e algo bem aleatório, as vezes crasha e as vezes não, mais alguém passando por isso?
> 
> Aqui não tive isso. Não sei se faz sentido também, porque a alteração que teve foi o som, aí daria debug sempre. Você ta usando a branch limpa? Client 13.16?

Acabei de atualizar tudo aqui, já testei também, problema resolvido 👊🏻
